### PR TITLE
Don't retype struct as primitive types in import.

### PR DESCRIPTION
**Pull request by sandreenko**
This change adds `COMPlus_JitNoStructRetyping` that prevents struct retyping for `call struct` and `return struct` cases. Currently, this retyping is happening in importer, we want to move it to lower.
The current retyping forbids later phases to do optimization with these values, for example, it affects code for inlined methods
```
struct nativeSizeStruct
{
  int a;
  int b;
}
nativeSizeStruct foo();
```
if we inline `foo()` we won't be able to promote or enregister fields of `nativeSizeStruct` because importer would access it as `LCL_FIELD long` and set `doNotEnreg` in `impFixupCallStructReturn`.

Why can't we do that right after inlining? Because other optimizations also need to know real types, for example, CSE and VN can't propagate a sequence like that:
```
ASG(nativeSizeStruct A, nativeSizeStruct B);
return LCL_FIELD long A;
```
but can do it if we don't do retyping.

<details>
<summary>Notes:</summary>
Note1: right after inlining is the right time to do retyping for methods that use return buffer, in that PR this is left in importer.

Note2: that change doesn't fix retyping in `impNormStructVal`, for cases like 
`methodWithStructArgs(foo(), foo()) ` we currently always create a local var for each struct local argument and retype it as a primitive type. That would be fixed in a separate PR.

Note3: that change doesn't fix retyping for limited struct promotion, when for a struct like:
struct PromotedStruct
{
   StructWithOneField a; <- we will change the type of this field from struct to int, because we don't have recursive struct promotion.
   int b;
}

struct StructWithOneField
{
  int a;
}
but that change will get us closer to it.

</details>

 So the phase in which we should retype struct types into ABI specific types is lowering. See details in https://github.com/dotnet/runtime/blob/master/docs/design/features/first-class-structs.md and https://github.com/dotnet/runtime/issues/1231.

The PR is done to have no diffs if `COMPlus_JitNoStructRetyping=0`, it helps to catch unwanted side-effects and makes merging safer. I have checked that there are no diffs on all windows platforms using altjit and framework assemblies. The change with the flag enabled was tested on SPMI collections, crossgen of framework libraries, and in a pri1 test run.

The overall design is simple: keep structs as struct until lowering, then do retyping for returns and calls, but insert `BITCAST` back to struct types to keep IR correct. Then teach the next phases (lsra, codegen) to work with the new struct nodes.

New cases of struct nodes force us to have struct handle on all trees (except the right side of ASG), so `gtGetStructHandleIfPresent` becomes more important.

Some initial diffs:
for the benchmark that motivated this change:
```
| ScalarFloatSinglethreadADT | \Core_Root_base\CoreRun.exe | 4.657 s | 0.0317 s | 0.0296 s | 4.650 s | 4.618 s | 4.710 s |  1.00 |             Base |     - |     - |     - |     272 B |
| ScalarFloatSinglethreadADT | \Core_Root_diff\CoreRun.exe | 1.157 s | 0.0083 s | 0.0078 s | 1.157 s | 1.145 s | 1.173 s |  0.25 |           Faster |     - |     - |     - |     272 B |
```
some other benchmarks are also winning, but not that significant:
```
| RegexRedux_1 | \Core_Root_base\CoreRun.exe | 603.4 ms | 96.37 ms | 110.97 ms | 612.5 ms | 424.6 ms | 762.8 ms |  1.00 |             Base |    0.00 |     - |     - |     - |   2.83 MB |
| RegexRedux_1 | \Core_Root_diff\CoreRun.exe | 431.6 ms |  7.89 ms |   6.99 ms | 432.3 ms | 418.2 ms | 443.7 ms |  0.76 |           Faster |    0.17 |     - |     - |     - |   2.83 MB |
```

code size changes for my small `StructABI\structreturn.cs` test, improvements happen when we inline the constructor method:

Overall, right now, it is a regression, I will start fixing them in the next change. Maybe I will push them to this PR or merge this PR with the flag disabled and fix the regressions in the next.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @CarolEidt @dotnet/jit-contrib, I think that is ready for the first round of review.

File: src/coreclr/src/jit/lsraxarch.cpp [hunk 1]
```diff
@@ -1364,7 +1367,18 @@ int LinearScan::BuildBlockStore(GenTreeBlk* blkNode)
...
+                srcRegMask = RBM_RSI;
+            }
+            else
+            {
+                assert(compiler->compNoReturnRetyping());
+                assert(src->OperIs(GT_BITCAST) && src->AsUnOp()->gtGetOp1()->IsCall());
+                // It could only happen if we the source was a struct call,
+                // we don't have a source address, we just need to generate one move.
+                srcRegMask = RBM_LNGRET;
+            }
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lsraxarch.cpp]: I think these changes add excess complexity and confusion to this method. I'd suggest to extract the case where the rhs is a call returning values in registers into a separate method.

File: src/coreclr/src/jit/lower.cpp [hunk 2]
```diff
@@ -3024,6 +3056,132 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
...
+#ifdef DEBUG
+        if (ret->TypeGet() != TYP_VOID)
+        {
+            if (varTypeIsStruct(ret->TypeGet()) != varTypeIsStruct(ret->gtGetOp1()->TypeGet()))
+            {
+
+                GenTree* retVal = ret->gtGetOp1();
+                if (retVal->OperIs(GT_LCL_VAR))
+                {
+                    // This could happen if we lied about op1 type during struct promotion,
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp]: I wouldn't describe this as lying about the type, but rather as returning the field instead of the full struct.
I'm not sure these lengthy comments add a lot of value. I think it would be better just to assert that we have compatible types.

File: src/coreclr/src/jit/lower.cpp [hunk 3]
```diff
@@ -3024,6 +3056,132 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
...
+        if (varTypeIsStruct(ret))
+        {
+            assert(comp->compNoReturnRetyping());
+            GenTree*  retVal           = ret->gtGetOp1();
+            var_types nativeReturnType = genActualType(comp->info.compRetNativeType);
+
+            ret->ChangeType(nativeReturnType);
+
+            assert(retVal->OperIs(GT_IND, GT_OBJ, GT_LCL_VAR, GT_LCL_FLD, GT_BITCAST, GT_CNS_INT, GT_SIMD));
+            switch (retVal->OperGet())
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp]: For the non-indirection cases, I think you should be able to handle them more uniformly - that is, if the register file type of the source is the same as the return type, then just retype the source. Otherwise add a `BITCAST`.

File: src/coreclr/src/jit/lowerxarch.cpp [hunk 4]
```diff
@@ -221,105 +221,121 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
     }
     else
     {
-        assert(src->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD));
-        src->SetContained();
-
-        if (src->OperIs(GT_IND))
+        if (src->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD))
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lowerxarch.cpp]: Why not make this an `else if`?

**Review by CarolEidt — COMMENTED** [review]
Overall looks good.
It's awesome to remove the retyping from the front-end, but I think that in Lowering when we have something that's returned in a single register, we should go ahead and retype. I think it would simplify things and would not lose information that the backend would need.

File: src/coreclr/src/jit/morph.cpp [hunk 5]
```diff
@@ -5835,11 +5839,11 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
...
 
-    CORINFO_FIELD_HANDLE symHnd          = tree->AsField()->gtFldHnd;
-    unsigned             fldOffset       = tree->AsField()->gtFldOffset;
-    GenTree*             objRef          = tree->AsField()->gtFldObj;
-    bool                 fieldMayOverlap = false;
-    bool                 objIsLocal      = false;
+    CORINFO_FIELD_HANDLE symHnd         = tree->AsField()->gtFldHnd;
+    unsigned             fldOffset      = tree->AsField()->gtFldOffset;
+    GenTree*             objRef         = tree->AsField()->gtFldObj;
+    bool                 noHndIfOverlap = false;
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp]: That part appeared to be incorrect, I have written a test case where we CSE wrong after that change, trying to find why.
After that I hope to change in the master branch or find how to keep the old behavior with the flag.

File: src/coreclr/src/jit/compiler.hpp [hunk 6]
```diff
@@ -956,6 +956,9 @@ inline GenTree* Compiler::gtNewOperNode(genTreeOps oper, var_types type, GenTree
     assert((GenTree::OperKind(oper) & GTK_EXOP) ==
            0); // Can't use this to construct any types that extend unary/binary operator.
     assert(op1 != nullptr || oper == GT_RETFILT || oper == GT_NOP || (oper == GT_RETURN && type == TYP_VOID));
+    // TODO seandree: Can't get address of some nodes, like call or simd, but the master branch does it
+    // and allows the later phases to eliminate these addr nodes. Fix such cases and uncoment that assert.
+    // assert(oper != GT_ADDR || !(op1->IsCall() || op1->OperIsSIMD()));
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/compiler.hpp]: Will do rewording, was trying to fix that but it was taking too long.

File: src/coreclr/src/jit/lclmorph.cpp [hunk 7]
```diff
@@ -521,6 +521,31 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                 PopValue();
                 break;
 
+            case GT_RETURN:
+                if (node->gtType != TYP_VOID)
+                {
+                    EscapeValue(TopValue(0), node);
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/lclmorph.cpp]: This is pre-existing logic, in `default ` we do:
```
                    EscapeValue(TopValue(0), node);
                    PopValue();
```

File: src/coreclr/src/jit/lclmorph.cpp [hunk 8]
```diff
@@ -521,6 +521,31 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
...
+                    {
+                        unsigned   lclNum = TopValue(0).LclNum();
+                        LclVarDsc* varDsc = m_compiler->lvaGetDesc(lclNum);
+                        if (varDsc->lvPromotedStruct() && (varDsc->lvFieldCnt != 1))
+                        {
+                            assert(m_compiler->compNoReturnRetyping() || varDsc->lvDoNotEnregister);
+                            // Its fields were accessed, so do not enregister it.
+                            // TODO seandree: it is possible to enregister that if we add a gather operation in codegen.
+                            // like struct s { bool a -> rax, bool b -> rdx, bool c -> rcx } -> rax,
+                            // return s -> gather(a, b, c) from their registers into rax.
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/lclmorph.cpp]: Yes, I will change the description.

> I would reword the above as "We can no longer promote or enregister this struct, since it is referenced as a whole." or something like that. It's not really that its fields are accessed.

I am not sure that we are talking about the same case. In the case that is covered by this code we have a struct that fits into a register, like `return struct A {bool a; bool b;}`, so we, in theory, can put it in a register. However, if we had accessed its fields, like `A.a = 1; A.b =2; ` then we can't put the whole struct in a register, we have to put it on the stack. The check for "Were the field accessed?" is implicit, if we had not accessed them we would not have promoted it.
And the special case covers structs with 1 field, in this case, we still can put it in a register.

[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/coreclr/src/jit/lclmorph.cpp]: Right. We could *in theory* put it into a register, but we don't currently have any support for accessing a multi-field struct as a whole. So perhaps "We can not promote this struct since it is referenced as a whole, and we don't currently have support for fully enregistering structs."
FWIW, enregistering (entire) structs is the second objective mentioned here: https://github.com/dotnet/runtime/blob/master/docs/design/features/first-class-structs.md#objectives (though perhaps it should be worded more clearly)

File: src/coreclr/src/jit/importer.cpp [hunk 9]
```diff
@@ -15209,6 +15237,18 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                    {
+                        if (op1->gtType == TYP_STRUCT)
+                        {
+                            op1->AsCall()->gtRetClsHnd = resolvedToken.hClass;
+                        }
+                        else
+                        {
+                            // We are doing unboxing, resolvedToken.hClass is available, but the result is byref
+                            // and gtRetClsHnd should not be used.
+                        }
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/importer.cpp]: I was sure that there were assert that was firing when I had `gtRetClsHnd ` on non-TYP_STRUCT nodes, but can't repro it now, probably it was my bad intermediate state, will fix.

File: src/coreclr/src/jit/lclmorph.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/lclmorph.cpp:540]: I've deleted the changes from lclmorph.cpp and moved that logic into `morph.cpp:fgMorphSmpOp:GT_RETURN` case, because there we have more information.

File: src/coreclr/src/jit/importer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/importer.cpp:15300]: I checked it again and found the reason: assert in `fgFixupStructReturn` is firing now on some platforms, 
https://github.com/dotnet/runtime/blob/73d817878aaec1e5056fd9ba94029dc79a4d5590/src/coreclr/src/jit/morph.cpp#L5011

It looks safe to delete it, but then we will have to collect diffs on all platforms and I think it is better to return the condition considering that all code under `compAllowReturnRetyping` will be deleted at the end.

File: src/coreclr/src/jit/lsraxarch.cpp (hunk 1 above)
Comment (sandreenko) [reply src/coreclr/src/jit/lsraxarch.cpp:1381]: Done, thanks.

File: src/coreclr/src/jit/morph.cpp [hunk 10]
```diff
@@ -9521,8 +9547,25 @@ GenTree* Compiler::fgMorphBlockOperand(GenTree* tree, var_types asgType, unsigne
         }
         else if (effectiveVal->TypeGet() != asgType)
         {
+#if !FEATURE_MULTIREG_RET
+            if (compNoReturnRetyping())
+            {
+                // TODO seandree: see if we can get rid of these cases or add a cast node.
```
[the remark this answers is not in the document]
Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp]: I have moved that logic to `fgMorphOneAsgBlockOp`.

**Comment by sandreenko** [comment]
I think this is ready for the second round, maybe expect `codegenxarch.cpp`.

No diffs if `JitAllowStructRetyping=1` (default), when the feature is enabled then it is a regression:
```
Crossgen CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies for  default jit
Summary of Code Size diffs:
(Lower is better)
Total bytes of diff: 79646 (0.24% of base)
```
but the biggest part of it due to `ASG(LCL_VAR struct with 1 field, call)` cases, that are part of https://github.com/dotnet/runtime/issues/34105, without these regressions it is a good improvement. 

I have checked SPMI/crossgen/pmi locally. I will kick appropriate CI testing this night.

File: src/coreclr/src/jit/gentree.cpp [hunk 11]
```diff
@@ -15099,31 +15099,42 @@ GenTree* Compiler::gtNewTempAssign(
     // struct types. We don't have a convenient way to do that for all SIMD temps, since some
     // internal trees use SIMD types that are not used by the input IL. In this case, we allow
     // a null type handle and derive the necessary information about the type from its varType.
-    CORINFO_CLASS_HANDLE structHnd = gtGetStructHandleIfPresent(val);
-    if (varTypeIsStruct(varDsc) && ((structHnd != NO_CLASS_HANDLE) || (varTypeIsSIMD(valTyp))))
+    CORINFO_CLASS_HANDLE valStructHnd = gtGetStructHandleIfPresent(val);
+    if (varTypeIsStruct(varDsc) && (valStructHnd == NO_CLASS_HANDLE) && !varTypeIsSIMD(valTyp))
+    {
+        // That is a very special case when we have lost classHandle from a LCL_FIELD node
```
Comment (sandreenko) [inline src/coreclr/src/jit/gentree.cpp:15105]: I am not happy with that solution, but the cost of having `fieldSeq` with information about overlapping fields was too high, both in terms of TP (checking the new flag in each access to it), in terms of memory consumption (make it 24 bytes instead of 16) and in implementation cost(there are dozens of places where we check for `NotAField`, many of them are very old and it was unclear how they should work with the new type.
Also, we don't often have overlapping fields, so it felt like doing I was spending too much on such a rare case. The frameworks showed -12 bytes of improvements from my fix that was an awful regression in TP and memory.

File: src/coreclr/tests/src/JIT/Directed/StructABI/StructWithOverlappingFields.cs [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+            Console.WriteLine("Second result: " + cseb0_2 + ", " + cse_uval_2 + ", " + cse_val_2 + ";");
+            Debug.Assert(cseb0_2 == 23);
+            Debug.Assert(cse_uval_2 == 7);
+            Debug.Assert(cse_val_2 == 52);
+        }
+
+        static int Main(string[] args)
+        {
+            TestClass(2);
```
Comment (sandreenko) [inline src/coreclr/tests/src/JIT/Directed/StructABI/StructWithOverlappingFields.cs:124]: The test uses a class and a struct to do the same logic, guess what case we optimize better. I should probably create an issue about that, but again, overlapping fields are rare.

**Comment by sandreenko** [comment]
> I would like to see us, perhaps in future, use BITCAST only where we actually require that the bits get moved to a different register file. We should be able to handle those cases much like the way multireg returns are handled. In fact, it's not clear why we need separate handling for that.

Hm, I think we are thinking about two different possible representations, with these changes we have the following new nodes in lowering:
`RETURN struct`,
`IND struct`,
`call struct`;
`LCL_VAR/LCL_FLD and similar`;
they could be combined in patterns like:
`RETURN struct (IND struct(ADDR ref (LCL_VAR struct)))`,
`RETURN struct(call struct)`,
`RETURN struct(SIMD SIMD8)`,
`STORE_OBJ(byref address, call struct)`.

I want to keep this struct representation visible even after lower, so for `STORE_OBJ(LCL_VAR struct, call struct)` do `STORE_OBJ(byref address, BITCAST<struct>(call long))`, the advantage is that we can see where we have structs and where we don't, bitcast nodes could do moves when the use and the def need different registers (like `call(call struct)`), they produce nothing if they don't need.

Another approach would be to produce the old LIR after lowering, meaning replace all such `STORE_OBJ` with native types `STORE_IND`, retype `LCL_VAR`, `CALL` into native types and try to avoid having any changes in `codegen` or `lsra`. We will need bitcast nodes only for `SIMD` and `LCL_FLD` nodes then.
I remember having issues with `STORE_IND byref`, but that could be fixed. Another issue with that would be `STORE_OBJ` destination. It is an address, but it knows the type, to which we are storing and it could be hidden deep in the dst tree. Finding it and changing it to a native type in lower, would be expensive, leaving it as-is will create strange trees like `STORE_IND int (IND struct(ADDR(LCL_VAR struct)), call int)` where dst type != src type.

**Comment by sandreenko** [comment]
ping @dotnet/jit-contrib

**Review by CarolEidt — COMMENTED** [review]
I'm still just a bit apprehensive about using `GT_BITCAST` for structs, but I think the idea is growing on me.
I'd like to better understand why the elimination of retyping is tied to `FEATURE_MULTIREG_RET` but otherwise it loosk good.

**Comment by tannergooding** [comment]
How will this impact `GT_HWINTRINSIC` and `GT_SIMD` which currently always lose the struct handle and retype to `TYP_SIMD8`, `TYP_SIMD12`, `TYP_SIMD16`, or `TYP_SIMD32` + a base type?

**Comment by CarolEidt** [comment]
> How will this impact GT_HWINTRINSIC and GT_SIMD which currently always lose the struct handle and retype to TYP_SIMD8, TYP_SIMD12, TYP_SIMD16, or TYP_SIMD32 + a base type?

I'll let @sandreenko reply to the specific question of how *this* work impacts those types (I think it's largely agnostic to those types). In future, however, now that we have a `ClassLayout` on `GenTreeJitIntrinsic` we should be able to retain the correct handle (and not have to look them up all over the place).

**Comment by AndyAyersMS** [comment]
> when the feature is enabled then it is a regression ... but the biggest part of it due to ASG(LCL_VAR struct with 1 field, call) cases, that are part of #34105, without these regressions it is a good improvement.

Can you say more about what the CS/CQ of this looks like once #34105 is fixed?

**Comment by sandreenko** [comment]
> > How will this impact GT_HWINTRINSIC and GT_SIMD which currently always lose the struct handle and retype to TYP_SIMD8, TYP_SIMD12, TYP_SIMD16, or TYP_SIMD32 + a base type?

> 
> I'll let @sandreenko reply to the specific question of how this work impacts those types (I think it's largely agnostic to those types). In future, however, now that we have a ClassLayout on GenTreeJitIntrinsic we should be able to retain the correct handle (and not have to look them up all over the place).

The change is agnostic to `GT_HWINTRINSIC`,  it touches `TYP_SIMD*` a bit, they are the difference between `varTypeIsStruct(type)` and `type == TYP_STRUCT` checks. As Carol has said, it is a move towards using the same logic for all `varTypeIsStruct(type)` like you can see in `Lowering::LowerRet(GenTreeUnOp* ret)` and some other cases. In future changes we will delete more special handling for `TYP_SIMD*`.

**Comment by sandreenko** [comment]
Ok, the tests are green now, there were a few new failures due to tail call changes.

The PR is ready for review.

@AndyAyersMS of course, I will repeat analysis for the diffs and post results here.

File: src/coreclr/src/jit/jitconfigvalues.h [hunk 13]
```diff
@@ -437,6 +437,14 @@ CONFIG_INTEGER(JitSaveFpLrWithCalleeSavedRegisters, W("JitSaveFpLrWithCalleeSave
 #endif // defined(TARGET_ARM64)
 #endif // DEBUG
 
+#if !FEATURE_MULTIREG_RET
+CONFIG_INTEGER(JitDoOldStructRetyping, W("JitDoOldStructRetyping"), 0) // Allow Jit to retype structs as primitive types
```
Comment (CarolEidt) [inline src/coreclr/src/jit/jitconfigvalues.h:441]: I like the new name - thanks :-)

File: src/coreclr/src/jit/gentree.cpp [hunk 14]
```diff
@@ -15137,6 +15137,14 @@ GenTree* Compiler::gtNewTempAssign(
             // and call returns. Lowering and Codegen will handle these.
             ok = true;
         }
+        else if (dstTyp == TYP_STRUCT && valTyp == TYP_INT)
```
Comment (CarolEidt) [inline src/coreclr/src/jit/gentree.cpp:15140]: nit: The expressions should be parenthesized.

File: src/coreclr/src/jit/gentree.cpp [hunk 15]
```diff
@@ -15137,6 +15137,14 @@ GenTree* Compiler::gtNewTempAssign(
...
+        {
+            // It could come from `ASG(struct, 0)` that was propogated to `RETURN struct(0)`,
```
Comment (CarolEidt) [inline src/coreclr/src/jit/gentree.cpp:15142]: Typo: propogated => propagated

File: src/coreclr/src/jit/gentree.cpp [hunk 16]
```diff
@@ -17277,6 +17309,9 @@ CORINFO_CLASS_HANDLE Compiler::gtGetStructHandleIfPresent(GenTree* tree)
 #endif
                 break;
         }
+        // TODO-1stClassStructs: add a check that `structHnd != NO_CLASS_HANDLE`,
```
Comment (CarolEidt) [inline src/coreclr/src/jit/gentree.cpp:17312]: Thanks for adding this `TODO`

File: src/coreclr/src/jit/lower.cpp [hunk 17]
```diff
@@ -2986,6 +3017,175 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
...
+    }
+}
+
+//----------------------------------------------------------------------------------------------
+// LowerCallStruct: Lowers a call node that returns a stuct.
+//
+// Arguments:
+//     call - The call node to lower.
+//
+// Note: it transforms the call's user.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp:3139]: It is a bit unfortunate that we have to transform the call's user. I wonder whether it would be feasible and reasonable to handle these when lowering the user. I think this is fine for now.

File: src/coreclr/src/jit/lower.cpp [hunk 18]
```diff
@@ -2986,6 +3017,175 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
...
+                // but `STORE` has to use the unwidened type.
+                user->ChangeType(returnType);
+                user->SetOper(GT_STOREIND);
+            }
+            break;
+
+            case GT_STOREIND:
+                assert(user->TypeIs(TYP_SIMD8, TYP_REF));
+                user->ChangeType(returnType);
+                user->SetOper(GT_STOREIND);
```
Comment (CarolEidt) [inline src/coreclr/src/jit/lower.cpp:3179]: Why do we need this? It's already `GT_STOREIND` right?

**Review by CarolEidt — APPROVED** [review]
Mostly comments and some non-blocking suggestions or questions.
Looks good!

File: src/coreclr/src/jit/lower.cpp (hunk 18 above)
Comment (sandreenko) [reply src/coreclr/src/jit/lower.cpp:3179]: Thanks for catching that, fixed.

File: src/coreclr/src/jit/lower.cpp (hunk 17 above)
Comment (sandreenko) [reply src/coreclr/src/jit/lower.cpp:3139]: You are right, we have agreed to do that in the user lowering but when I did so I saw that:
1) we have to process call in `LowerCall` when it is unused, so compilation time is spent to find a user in any case;
2) the processing goes into 2 functions (in addition to `LowerCall`) `LowerBlockStore, LowerStoreIndir` and they are platform-specific, so the logic to find these cases should be duplicated for different platforms;
3) it takes more code and operations to find these cases from the user, because each struct call has its user that should be modified, but not all struct stores have such calls as they src.

so after a while, I changed it to process it in one place, I think it is more visible.

**Comment by sandreenko** [comment]
@AndyAyersMS 
examples of some improvements (I will write a complete report when we enable this by default):

<details>

<summary>the first type, -580 (-40.93% of base) : Microsoft.CodeAnalysis.CSharp.dasm - BinopEasyOut:TypeToIndex(TypeSymbol):Nullable`1</summary>

This is the first type of improvements.
This method returns a `struct<8> { 0x0 bool, 0x4 int }` and has 30 `return` instructions, so before the change we were retyping `Return struct(LCL_VAR struct V01)` to `Return long(LCL_FLD long V01)` for all of them. It was forbiding independent struct promotion of all involved `LCL_VAR`. Now we keep them as structs, create a merge return `LCL_VAR`, but we copy results from `LCL_VAR`s that are promoted independently, so instead of that:

```
N003 (  3,  4) [000707] -A------R---              *  ASG       long  
N002 (  1,  1) [000706] D------N----              +--*  LCL_VAR   long   V35 tmp32        
N001 (  3,  4) [000078] ------------              \--*  LCL_FLD   long   V11 tmp8         [+0]
                                                  \--*    bool   V11.hasValue (offs=0x00) -> V52 tmp49        
                                                  \--*    int    V11.value (offs=0x04) -> V53 tmp50  
```

where optimizations of both `V35` and `V11` are blocked, we have:

```
N007 ( 16, 12) [000763] -A----------              *  COMMA     void  
N003 (  9,  7) [000759] -A------R---              +--*  ASG       bool  
N002 (  4,  3) [000757] D------N----              |  +--*  LCL_VAR   bool   V94 tmp91        
N001 (  4,  3) [000758] -------N----              |  \--*  LCL_VAR   bool   V52 tmp49        
N006 (  7,  5) [000762] -A------R---              \--*  ASG       int   
N005 (  3,  2) [000760] D------N----                 +--*  LCL_VAR   int    V95 tmp92        
N004 (  3,  2) [000761] -------N----                 \--*  LCL_VAR   int    V53 tmp50
```

Then we apply CSE and propagation optimizations that benefit from `bool` lclVars and propagate that field assignment well. As a result, we have 58 `Removing tree [TREE_ID] in [BB_ID] as useless` (0 in base) and then allocate the rest to registers instead of memory.
ASM before:

```
IN0011: 000054 mov      byte  ptr [V52 rsp+C0H], 0
IN0012: 00005C xor      eax, eax
IN0013: 00005E mov      dword ptr [V53 rsp+C4H], eax
IN0014: 000065 mov      dword ptr [V53 rsp+C4H], 6
IN0015: 000070 mov      byte  ptr [V52 rsp+C0H], 1
IN0016: 000078 mov      rax, qword ptr [V11 rsp+C0H]
```

asm after:

```
IN0011: 000043 mov      eax, 6
IN0012: 000048 mov      ecx, 1
IN0013: 00004D mov      byte  ptr [V94 rsp+20H], cl                                           
IN0014: 000051 mov      dword ptr [V95 rsp+24H], eax
```

</details>

<details>

<summary>the second type, -203 (-20.86% of base) : Newtonsoft.Json.dasm - DefaultContractResolver:CreatePropertyFromConstructorParameter(JsonProperty,ParameterInfo):JsonProperty:this</summary>

This is the second type of improvements, that was the initial motivation for that work (from SIMD bench).

We are inlining a method that returns a small struct:

```
    [000149] -AC---------              *  ASG       struct (copy)                                                   
    [000147] D------N----              +--*  LCL_VAR   struct<System.Nullable`1[Boolean], 2> V06 loc3         
    [000146] --C---------              \--*  RET_EXPR  struct(inl return from call [000145])
```

and before we were transforming it to:

```
    [000150] -AC---------              *  ASG       short 
    [000149] ------------              +--*  IND       short 
    [000148] ------------              |  \--*  ADDR      byref 
    [000147] -------N----              |     \--*  LCL_VAR   struct<System.Nullable`1[Boolean], 2> V06 loc3         
    [000146] --C---------              \--*  RET_EXPR  int   (inl return from call [000145])
```

that was blocking `V06` struct promotion and enregestering, now it does not happen.
so when we copy these values instead of two moves (from a memory loc to a reg and from the reg to another mem loc) we have one reg to reg move:

```
movzx    rax, byte  ptr [V54 rsp+90H]
mov      byte  ptr [V68 rsp+60H], al
```

after:

```
movzx    rcx, dl
```

</details>


<details>

<summary>the third type, -13 (-81.25% of base) : System.Private.CoreLib.dasm - ValueTuple:Create():ValueTuple</summary>

That is a funny one, for IL like:

```
Importing BB01 (PC=000) of 'System.ValueTuple:Create():System.ValueTuple'
    [ 0]   0 (0x000) ldloca.s 0
    [ 1]   2 (0x002) initobj 020001C9
```

we were generating 

```
***** BB01
STMT00000 (IL 0x000...0x003)
               [000003] IA----------              *  ASG       struct (init)
               [000000] D------N----              +--*  LCL_VAR   struct<System.ValueTuple, 1> V00 loc0         
               [000002] ------------              \--*  CNS_INT   int    0

***** BB01
STMT00001 (IL 0x008...0x009)
               [000005] ------------              *  RETURN    int   
               [000004] ------------              \--*  LCL_FLD   byte   V00 loc0         [+0]
```

now it is 

```
N002 (  2,  2) [000005] ------------              *  RETURN    struct
N001 (  1,  1) [000004] ------------              \--*  CNS_INT   int    0
```

(copy propagation for structs was unblocked in one of the preparation PRs).
so instead of

```
IN0004: 000000 push     rax
IN0001: 000001 xor      eax, eax
IN0002: 000003 mov      byte  ptr [V00 rsp], al
IN0003: 000006 movsx    rax, byte  ptr [V00 rsp]
IN0005: 00000B add      rsp, 8
IN0006: 00000F ret
```

we now have

```
IN0001: 000000 xor      eax, eax
IN0002: 000002 ret
```

when we create a `ValueTuple` (wonder if it is popular, like std::map<T,S> in C++).
</details>

After #34105 and #11413 are fixed we will probably see some other patterns as well instead of the current regressions.

**Comment by AndyAyersMS** [comment]
> examples of some improvements

@sandreenko looks good. I would expect `Nullable<int>` to be one of the things that really benefits here.  You might want to explore add this as an instantiation type in PMI to see more broadly what happens. You should also see the `Range` type (struct of two ints) benefit, see notes over in #11848.

**Comment by sandreenko** [comment]
> @sandreenko looks good. I would expect Nullable<int> to be one of the things that really benefits here. You might want to explore add this as an instantiation type in PMI to see more broadly what happens. You should also see the Range type (struct of two ints) benefit, see notes over in #11848.

yes, all returns and calls that return structs in a register should benefit.

I have added `Type[] typesToTry = typeof(int?)` and got additional  216 methods improved. The full logs are attached.
I have looked at a few regressed methods with `int?` and they are in the expected form of  `ASG(LCL_VAR struct with 1 field, call struct) ` where we now block independent promotion of the struct.
[pmiDiffsWithNullInt.txt](https://github.com/dotnet/runtime/files/4578473/pmiDiffsWithNullInt.txt)
[crossgenDiffs.txt](https://github.com/dotnet/runtime/files/4578477/crossgenDiffs.txt)
[pmiDiffs.txt](https://github.com/dotnet/runtime/files/4578478/pmiDiffs.txt)
